main.go

```diff
@@ -148,24 +148,45 @@ func getConnWithRetry(flags *flagStorage) (c *gcsx.Connection, err error) {
 	return
 }
 
-func createStorageHandle(flags *flagStorage) (storageHandle storage.StorageHandle, err error) {
-	tokenSrc, err := auth.GetTokenSource(context.Background(), flags.KeyFile, flags.TokenUrl, true)
-	if err != nil {
-		err = fmt.Errorf("get token source: %w", err)
-		return
+func handleCustomEndpoint(flags *flagStorage, config *storage.StorageClientConfig) (err error) {
+	var tokenSrc oauth2.TokenSource
+	var opts []option.ClientOption
+	if flags.Endpoint == nil || flags.Endpoint.Hostname() == "storage.googleapis.com" {
+		tokenSrc, err = auth.GetTokenSource(context.Background(), flags.KeyFile, flags.TokenUrl, flags.ReuseTokenFromUrl)
+		if err != nil {
+			err = fmt.Errorf("GetTokenSource: %w", err)
+			return
+		}
+	} else {
+
+		// Do not use OAuth with non-Google hosts.
+		tokenSrc = oauth2.StaticTokenSource(&oauth2.Token{})
+		opts = append(opts, option.WithEndpoint(flags.Endpoint.String()))
 	}
 
+	config.TokenSrc = tokenSrc
+	config.ClientOptions = opts
```

Comment:
But that type is publicly exposed.

Anyway, the reason behind adding this instead of Endpoint url is we can pass multiple options from here. For e.g. we are passing WithJSONReads in this PR-https://github.com/GoogleCloudPlatform/gcsfuse/pull/1265 (incomplete). In future, also we don't have to pass different parameter for such options.

Reply:
Got it, in that case WithHTTPClient should also be passed in option.ClientOption instead of creating that in NewStorageHandle

Reply:
Yeah, that will make the NewStorageHandle method clean, also simplify the StorageClientConfig object.

Reply:
Create client_helper to separate all client creation logic at one place.